cmd/litcli/ln.go

```diff
@@ -0,0 +1,637 @@
+			FeeRateSatPerVbyte: uint32(c.Uint64("sat_per_vbyte")),
+			PushSat:            c.Int64("push_amt"),
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("error funding channel: %w", err)
+	}
+
+	printJSON(resp)
+
+	return nil
+}
+
+var (
+	assetIDFlag = cli.StringFlag{
+		Name: "asset_id",
+		Usage: "the asset ID of the asset to use when sending " +
+			"payments with assets",
+	}
+
+	assetAmountFlag = cli.Uint64Flag{
+		Name: "asset_amount",
+		Usage: "the amount of the asset to send in the asset keysend " +
+			"payment",
+	}
+
+	rfqPeerPubKeyFlag = cli.StringFlag{
+		Name: "rfq_peer_pubkey",
+		Usage: "(optional) the public key of the peer to ask for a " +
+			"quote when converting from assets to sats; must be " +
+			"set if there are multiple channels with the same " +
+			"asset ID present",
+	}
+)
+
+// resultStreamWrapper is a wrapper around the SendPaymentClient stream that
+// implements the generic PaymentResultStream interface.
+type resultStreamWrapper struct {
+	amountMsat int64
+	stream     tchrpc.TaprootAssetChannels_SendPaymentClient
+}
+
+// Recv receives the next payment result from the stream.
+//
+// NOTE: This method is part of the PaymentResultStream interface.
+func (w *resultStreamWrapper) Recv() (*lnrpc.Payment, error) {
+	resp, err := w.stream.Recv()
+	if err != nil {
+		return nil, err
+	}
+
+	res := resp.Result
+	switch r := res.(type) {
+	// The very first response might be an accepted sell order, which we
+	// just print out.
+	case *tchrpc.SendPaymentResponse_AcceptedSellOrder:
+		quote := r.AcceptedSellOrder
+		rpcRate := quote.BidAssetRate
+		rate, err := rfqrpc.UnmarshalFixedPoint(rpcRate)
+		if err != nil {
+			return nil, fmt.Errorf("unable to unmarshal fixed "+
+				"point: %w", err)
+		}
+
+		amountMsat := lnwire.MilliSatoshi(w.amountMsat)
+		milliSatsFP := rfqmath.MilliSatoshiToUnits(amountMsat, *rate)
+		numUnits := milliSatsFP.ScaleTo(0).ToUint64()
+
+		// If the calculated number of units is 0 then the asset rate
+		// was not sufficient to represent the value of this payment.
+		if numUnits == 0 {
+			// We will calculate the minimum amount that can be
+			// effectively sent with this asset by calculating the
+			// value of a single asset unit, based on the provided
+			// asset rate.
+
+			// We create the single unit.
+			unit := rfqmath.FixedPointFromUint64[rfqmath.BigInt](
+				1, 0,
+			)
+
+			// We derive the minimum amount.
+			minAmt := rfqmath.UnitsToMilliSatoshi(unit, *rate)
+
+			// We return the error to the user.
+			return nil, fmt.Errorf("smallest payment with asset "+
+				"rate %v is %v, cannot send %v",
+				rate.ToUint64(), minAmt, amountMsat)
+		}
+
+		msatPerUnit := uint64(w.amountMsat) / numUnits
+
+		fmt.Printf("Got quote for %v asset units at %v msat/unit from "+
+			"peer %s with SCID %d\n", numUnits, msatPerUnit,
+			quote.Peer, quote.Scid)
+
+		resp, err = w.stream.Recv()
+		if err != nil {
+			return nil, err
+		}
+
+		return resp.GetPaymentResult(), nil
```

Comment:
Could maybe be worth `nil` checking the result of `resp.GetPaymentResult()` before returning here?